src/schema/sources.ts

```diff
@@ -1213,6 +1213,45 @@ export const ensureSourcePermissions = async (
   throw new ForbiddenError('Access denied!');
 };
 
+export const ensureUserSourceExists = async (userId: string, con: DataSource) =>
+  con.transaction(async (entityManager) => {
+    const user = await entityManager
+      .getRepository(User)
+      .findOneByOrFail({ id: userId });
+
+    await entityManager.getRepository(SourceUser).upsert(
+      {
+        id: user.id,
+        userId: user.id,
+        handle: user.username,
+        name: user.username,
+        type: SourceType.User,
+        private: false,
+        flags: {
+          publicThreshold:
+            user.reputation >= REPUTATION_THRESHOLD && !user.flags.vordr,
+          vordr: user.flags.vordr ?? false,
+        },
+      },
+      ['id'],
+    );
+
+    const sourceMemberExists = await entityManager
+      .getRepository(SourceMember)
+      .findOne({
+        where: { sourceId: user.id, userId: user.id },
+      });
+
+    if (!sourceMemberExists) {
+      await entityManager.getRepository(SourceMember).insert({
```

Comment:
As https://github.com/dailydotdev/daily-api/pull/2881/files#r2175178994 this could still be race condition, better to use on conflict do nothing with query builder.

Reply:
Also, just remembered, we technically would need to add ContentPreferenceSource as well as we do for joinSource because of backward compatibility.

Reply:
Convert to qb with do nothing on conflict, also added ContentPreferenceSource upsert